addons/xterm-addon-image/src/ImageRenderer.ts

```diff
@@ -0,0 +1,352 @@
+/**
+ * Copyright (c) 2020 The xterm.js authors. All rights reserved.
+ * @license MIT
+ */
+
+import { toRGBA8888 } from 'sixel/lib/Colors';
+import { IDisposable } from 'xterm';
+import { ICellSize, ICoreTerminal, IImageSpec, IRenderDimensions, IRenderService } from './Types';
+
+
+const PLACEHOLDER_LENGTH = 4096;
+const PLACEHOLDER_HEIGHT = 24;
+
+/**
+ * ImageRenderer - terminal frontend extension:
+ * - provide primitives for canvas, ImageData, Bitmap (static)
+ * - add canvas layer to DOM (browser only for now)
+ * - draw image tiles onRender
+ */
+export class ImageRenderer implements IDisposable {
+  public canvas: HTMLCanvasElement | undefined;
+  private _ctx: CanvasRenderingContext2D | null | undefined;
+  private _placeholder: HTMLCanvasElement | undefined;
+  private _placeholderBitmap: ImageBitmap | undefined;
+  private _optionsRefresh: IDisposable | undefined;
+  private _oldOpen: ((parent: HTMLElement) => void) | undefined;
+  private _renderService: IRenderService | undefined;
+  private _oldSetRenderer: ((renderer: any) => void) | undefined;
+
+  // drawing primitive - canvas
+  public static createCanvas(width: number, height: number): HTMLCanvasElement {
+    const canvas = document.createElement('canvas');
+    canvas.width = width | 0;
+    canvas.height = height | 0;
+    return canvas;
+  }
+
+  // drawing primitive - ImageData with optional buffer
+  public static createImageData(ctx: CanvasRenderingContext2D, width: number, height: number, buffer?: ArrayBuffer): ImageData {
+    if (typeof ImageData !== 'function') {
+      const imgData = ctx.createImageData(width, height);
+      if (buffer) {
+        imgData.data.set(new Uint8ClampedArray(buffer, 0, width * height * 4));
+      }
+      return imgData;
+    }
+    return buffer
+      ? new ImageData(new Uint8ClampedArray(buffer, 0, width * height * 4), width, height)
+      : new ImageData(width, height);
+  }
+
+  // drawing primitive - ImageBitmap
+  public static createImageBitmap(img: ImageBitmapSource): Promise<ImageBitmap | undefined> {
+    if (typeof createImageBitmap !== 'function') {
+      return new Promise(res => res(undefined));
```

Comment:
```suggestion
      return Promise.resolve(undefined);
```

Can use async to return undefined if it works in static functions